crates/forge/bin/cmd/test/mod.rs

```diff
@@ -311,20 +322,81 @@ impl TestArgs {
             .enable_isolation(evm_opts.isolate)
             .build(project_root, &output, env, evm_opts)?;
 
-        if let Some(debug_test_pattern) = &self.debug {
+        if should_debug && should_flamegraph {
+            eyre::bail!("Cannot specify both --debug and --flamegraph.");
+        }
+
+        let mut set_test_pattern = |new_test_pattern: &Regex, flag_name| {
             let test_pattern = &mut filter.args_mut().test_pattern;
             if test_pattern.is_some() {
                 eyre::bail!(
-                    "Cannot specify both --debug and --match-test. \
-                     Use --match-contract and --match-path to further limit the search instead."
+                    "Cannot specify both --{} and --match-test. \
+                     Use --match-contract and --match-path to further limit the search instead.",
+                    flag_name
                 );
             }
-            *test_pattern = Some(debug_test_pattern.clone());
+            *test_pattern = Some(new_test_pattern.clone());
+            Ok(())
+        };
+
+        if let Some(debug_test_pattern) = &self.debug {
+            set_test_pattern(debug_test_pattern, "debug")?;
+        }
+
+        if let Some(flamegraph_test_pattern) = &self.flamegraph {
+            set_test_pattern(flamegraph_test_pattern, "flamegraph")?;
         }
 
         let libraries = runner.libraries.clone();
         let outcome = self.run_tests(runner, config, verbosity, &filter, &output).await?;
 
+        if should_flamegraph {
+            let (_, test_result) = outcome
+                .results
+                .iter()
+                .find(|(_, r)| !r.test_results.is_empty())
+                .map(|(_, r)| (r, r.test_results.values().next().unwrap()))
+                .unwrap();
+
+            let arena = test_result
+                .traces
+                .iter()
+                .find_map(
+                    |(kind, arena)| {
+                        if *kind == TraceKind::Execution {
+                            Some(arena)
+                        } else {
+                            None
+                        }
+                    },
+                )
+                .unwrap();
+
+            let decoder = outcome.last_run_decoder.as_ref().unwrap();
+            decoder.prefetch_signatures(arena.nodes()).await;
+
+            let mut folded_stack_lines = render_trace_arena(arena, decoder).await?.1;
+            folded_stack_lines.reverse();
+
+            let file_name = "flamegraph.svg";
+            let writer = std::fs::File::create(file_name).unwrap();
+
+            let mut options = inferno::flamegraph::Options::default();
+            options.flame_chart = true;
```

Comment:
The inferno crate provides some options. Some of them could be passed on to user like choosing between flamegraph and flamechart. 

Just realised that adding these flags on the test subcommand is not a good idea. Separate subcommand like `forge flamegraph` makes more sense? (I did this in test subcommand to prevent code duplication)